### PR TITLE
Add dockerfiles for garage

### PR DESCRIPTION
**Pull request by gonzaiva**
Two versions were added:
- Release: image based on the latest stable release from garage
- Local: image built on top of the release image with local changes of
the garage repository

A makefile was added as well to build and run the corresponding
containers, as well as a README.md with the documentation for the docker
images.

### CODE REVIEW

File: docker/README.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
+# Garage Docker
```
Comment (gonzaiva) [inline docker/README.md]: I forgot to add a pre-requisites section, mainly for docker-ce and docker-compose.

File: .gitignore [hunk 2]
```diff
@@ -32,6 +32,7 @@ iterate.dat
 src/
 .settings
 .pods
+docker/mjkey.txt
```
Comment (ryanjulian) [inline .gitignore]: don't copy this into the repo. retrieve it from wherever it lives on the filesystem and then pass it as an environment variable. see Dockerfile.ci for an example.

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,34 @@
+.PHONY: build-garage-release run-garage-release
+
+GARAGE_RELEASE="v2018.10.0"
+
+build-garage-release: docker/Dockerfile.release docker/mjkey.txt
```
Comment (ryanjulian) [inline Makefile]: nah this makes no sense. there is only one set of canonical Dockerfiles with a single set of names.

the ones in the release branch are compliant with the release. the ones on the master branch are compliant with master.

Dockerfiles are code, just like any other Python file. we don't have -release and -local versions of logger.py, do we?

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	docker build \
+		-f docker/Dockerfile.release \
+		--build-arg RELEASE="${GARAGE_RELEASE}" \
+		-t rlworkgroup/garage:release .
```
Comment (ryanjulian) [inline Makefile]: tags (garage:release) are for specifying versions, not variants. if you want an image variant, you should name it with a dash.

for instance, the ci version of this Dockerfile is garage-ci

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+build-garage-local: docker/Dockerfile.local
+	docker build \
+		-f docker/Dockerfile.local \
+		-t rlworkgroup/garage:local .
```
Comment (ryanjulian) [inline Makefile]: tags are for versions, not variants.

File: docker/README.md (hunk 1 above)
Comment (ryanjulian) [inline docker/README.md]: I think this is more appropriately called "Using Docker to run garage"

File: docker/README.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+Currently there's three versions of garage containers:
+  - ci: used for the continuous integration of garage in TravisCI and not
+    intended for final users.
+  - release: it's a published image based on the latest official release of
```
Comment (ryanjulian) [inline docker/README.md]: this should just live in the release branch

File: docker/README.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    garage.
+  - local: it's an image built on top of the release with a copy of the files
```
Comment (ryanjulian) [inline docker/README.md]: this is not called .local, it's just called the Dockerfile. 

whereas we put special Dockerfiles in docker/, the canonical Dockerfile usually goes in the repository root. We might move all Dockerfiles to the repo root--not sure yet. That's out of scope for this PR.

File: docker/README.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+```
+docker run --rm rlworkgroup/garage:release RUN_CMD="python examples/tf/trpo_cartpole.py"
+```
+This image does not come with a MuJoCo key, so it's not possible to run MuJuCo
+environments with it. To enable MuJoCo, you need to build a local image and
+set a valid key in the image.
+
+If you want to build the release image, follow this steps:
+  1. Clone this repository
+  2. Copy a valid MuJoCo key in `docker/mjkey.txt` inside the cloned repository
```
Comment (ryanjulian) [inline docker/README.md]: No. See how we do it in Dockerfile.ci. We do that for a reason.

File: docker/README.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+```
+
+This previous command adds a volume from the data folder inside your cloned
+garage repository to the garage-release container, so any experiment results
+ran in the container will be saved in the data folder inside your cloned
+repository.
+
+## Local image
+
+The local image is built on top of the release image and it includes the
```
Comment (ryanjulian) [inline docker/README.md]: This doesn't make any sense to me. What if I want to use the latest garage, not the release?

File: docker/README.md [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+changes in your cloned repository as well as configuring the MuJoCo key
+(optionally).
+
+If you want to build the local image, follow this steps:
+  1. Clone this repository
+  2. Copy a valid MuJoCo key in `docker/mjkey.txt` inside the cloned repository
```
Comment (ryanjulian) [inline docker/README.md]: No. Use environment variables.

File: docker/docker-entrypoint-release.sh [hunk 11]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
+# Dockerfile entrypoint
+set -e
+
+source ~/.bashrc
+
+# Activate conda environment
+source activate garage
```
Comment (ryanjulian) [inline docker/docker-entrypoint-release.sh]: How does this differ from the CI entrypoint?

File: docker/Dockerfile.release [hunk 12]
```diff
@@ -0,0 +1,108 @@
+# NOTICE: To keep consistency across this docker file, scripts/setup_linux.sh
+# and scripts/setup_macos.sh, if there's any changes applied to this file,
+# specially regarding the installation of dependencies, apply those same
+# changes to the mentioned files.
+FROM ubuntu:16.04
```
Comment (ryanjulian) [inline docker/Dockerfile.release]: This image uses a fake display, meaning I can't ever see the training visualization. Why not have fake display, docker-opengl, and nvidia-docker versions of this?

The canonical one would be docker-opengl, since that will run on any machine interactively.

File: docker/README.md [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    in your local repository.
+
+## Release image
+
+The release image is published to be downloaded without the need of having
+a local repository. The command to run in the corresponding container can be
+specified through the argument RUN_CMD. To run an example launcher in the
+container, execute:
+```
+docker run --rm rlworkgroup/garage:release RUN_CMD="python examples/tf/trpo_cartpole.py"
```
Comment (ryanjulian) [inline docker/README.md]: why do i need RUN_CMD? I can just put the command right on the terminal and it will be `$@` in the entrypoint

**Comment by gonzaiva** [comment]
Hi guys, I'm back and I'll be reviewing your comments today.

File: .gitignore (hunk 2 above)
Comment (gonzaiva) [reply .gitignore]: Done, now the mjkey is read from an environment variable instead.

File: Makefile (hunk 3 above)
Comment (gonzaiva) [reply Makefile]: Got it! Once this PR is approved, I will back port this commit to the 2018.10 release to build the corresponding version images.

File: Makefile (hunk 4 above)
Comment (gonzaiva) [reply Makefile]: I followed this nomenclature in the new changes

File: docker/README.md (hunk 1 above)
Comment (gonzaiva) [reply docker/README.md]: Done

File: docker/README.md (hunk 6 above)
Comment (gonzaiva) [reply docker/README.md]: Once approved, I will back port this PR to the release branch. In the meantime, I have removed the image named rlworkgroup/garage:release.

File: docker/README.md (hunk 8 above)
Comment (gonzaiva) [reply docker/README.md]: Done

File: docker/README.md (hunk 9 above)
Comment (gonzaiva) [reply docker/README.md]: Got it! This PR will create the images for latest changes in master, and each release branch will have their corresponding images based on the release number.

File: docker/README.md (hunk 10 above)
Comment (gonzaiva) [reply docker/README.md]: Done

File: docker/docker-entrypoint-release.sh (hunk 11 above)
Comment (gonzaiva) [reply docker/docker-entrypoint-release.sh]: I removed this file and with some light modifications I'm using the CI entry point.

File: docker/Dockerfile.release (hunk 12 above)
Comment (gonzaiva) [reply docker/Dockerfile.release]: I could generate images for the fake display (I named it headless) and nvidia versions. However, I couldn't find/create a proper opengl image that works with ubuntu 16.04. I'm taking a look at which packages the nvidia-opengl uses to copy the opengl parts only, but I can submit this version in a next PR.

File: docker/README.md [hunk 14]
```diff
@@ -0,0 +1,81 @@
+# Garage Docker
+
+Currently there's three versions of garage containers:
```
[the remark this answers is not in the document]
Comment (gonzaiva) [reply docker/README.md]: Done

File: docker/README.md (hunk 13 above)
Comment (gonzaiva) [reply docker/README.md]: Fixed!

File: Makefile (hunk 3 above)
Comment (ryanjulian) [reply Makefile]: I honestly don't think that's necessary. 2018.10 release only guarantees bug fixes, not new features.

File: .travis.yml [hunk 15]
```diff
@@ -75,6 +76,10 @@ after_success:
       echo "${DOCKER_API_KEY}" | docker login -u "${DOCKER_USERNAME}" --password-stdin
       docker tag "${tag}" rlworkgroup/garage-ci:latest
       docker push rlworkgroup/garage-ci
+      make build-garage-headless
+      docker push rlworkgroup/garage-headless
+      make build-garage-nvidia
```
Comment (ryanjulian) [inline .travis.yml]: No no no. Do not build the nvidia dockerfile in the CI.

This won't work because the CI has no graphics card.

Also the purpose of this part of the CI script is not to maintain the Docker hub. It's to keep a cache of the CI docker image. Please remove this.

File: docker/Dockerfile [hunk 16]
```diff
@@ -105,7 +106,14 @@ RUN curl -L https://codeclimate.com/downloads/test-reporter/test-reporter-latest
 
 # Add code stub last
 COPY . /root/code/garage
+# Prevent garage from exiting on config
+RUN cp /root/code/garage/garage/config_personal_template.py \
+  /root/code/garage/garage/config_personal.py
 RUN ["/bin/bash", "-c", "source activate garage && pip install -e /root/code/garage && source deactivate garage"]
 
+# Specifies which docker image flavor we're dealing with
+ARG IMAGE_TYPE
```
Comment (ryanjulian) [inline docker/Dockerfile]: this is strange. why does the Dockerfile need to know this?

File: docker/Dockerfile [hunk 17]
```diff
@@ -105,7 +106,14 @@ RUN curl -L https://codeclimate.com/downloads/test-reporter/test-reporter-latest
 
 # Add code stub last
 COPY . /root/code/garage
+# Prevent garage from exiting on config
+RUN cp /root/code/garage/garage/config_personal_template.py \
```
Comment (ryanjulian) [inline docker/Dockerfile]: Why is this necessary here but not in the headless version?

File: Makefile [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+	xhost +local:docker
+	@ docker run \
+		-it \
+		--rm \
+		--runtime=nvidia \
+		-v /tmp/.X11-unix:/tmp/.X11-unix \
+		-v $(shell pwd)/data:/root/code/garage/data \
+		-e DISPLAY=$(DISPLAY) \
+		-e QT_X11_NO_MITSHM=1 \
+		-e MJKEY="$(MJKEY)" \
```
Comment (ryanjulian) [inline Makefile]: suggestion:
if the environment variable MJKEY is empty, attempt to read it from the disk (~/.mujoco/mjkey.txt) into the environment variable and produce a message saying you automatically detected the key at that location.

File: docker/Dockerfile [hunk 19]
```diff
@@ -84,7 +85,7 @@ COPY environment.yml /root/code/garage/environment.yml
 # MAKE SURE TO DELETE MJKEY.TXT SO THAT WE DON'T PUBLISH THE KEY
 # NOTE: You MUST create and delete the mjkey.txt within the same RUN command
 ARG MJKEY
-RUN echo "${MJKEY}" > /root/.mujoco/mjkey.txt && \
+RUN echo -e "${MJKEY}" > /root/.mujoco/mjkey.txt && \
```
Comment (ryanjulian) [inline docker/Dockerfile]: why is -e necessary?

File: docker/README.md [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+docker run -it --rm rlworkgroup/garage-headless python examples/tf/trpo_cartpole.py
+```
+
+This image does not come with a MuJoCo key, so if you want to run environments
+using MuJoCo, the environment variable `MJKEY` has to be set with the
+corresponding content of the key file in your file system:
+```
+# New line characters in MuJoCo key files (they're in DOS format) are
+# replaced by '\n' to pass the file as a command argument.
+export MJKEY="$(sed ':a;N;$!ba;s/\r\r//g;s/\n/\\n/g' <your_key_path>)"
```
Comment (ryanjulian) [inline docker/README.md]: Do not instruct users to export their MJKEY. This is a security risk.

It's extremely easy to include your MJKEY as an argument

MJKEY="$(cat ~/.mujoco/mjkey.txt)"

File: docker/README.md [hunk 21]
```diff
@@ -0,0 +1,79 @@
+# Using Docker to run garage
+
+Currently there are three types of garage images:
+  - ci: used for the continuous integration of garage in TravisCI and not
```
Comment (ryanjulian) [inline docker/README.md]: you appear to only have CI and headless images? also it's not clear how CI differs from headless

File: docker/README.md [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+    intended for final users.
```
Comment (ryanjulian) [inline docker/README.md]: what is a final user?

File: docker/README.md [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+  - headless: garage without environment visualization.
+  - NVIDIA: garage with environment visualization using an NVIDIA graphics
```
Comment (ryanjulian) [inline docker/README.md]: lower-case nvidia

File: docker/README.md [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+    card.
+
```
Comment (ryanjulian) [inline docker/README.md]: what about users who want to run interactively but without an nvidia card? see docker-opengl

File: docker/README.md [hunk 25]
```diff
@@ -0,0 +1,79 @@
+# Using Docker to run garage
+
+Currently there are three types of garage images:
```
Comment (ryanjulian) [inline docker/README.md]: it's not clear how headless and CI differ

File: docker/README.md [hunk 26]
```diff
@@ -0,0 +1,79 @@
...
+  -it \
+  --rm \
+  -e MJKEY="${MJKEY}" \
+  rlworkgroup/garage-headless python examples/theano/trpo_swimmer.py
+```
+
+For every release of garage there's a version of this container, which you can
+specify through the release number. For example:
+```
+docker run --rm rlworkgroup/garage-headless:2018.10 python examples/tf/trpo_cartpole.py
```
Comment (ryanjulian) [inline docker/README.md]: nah

File: docker/docker-entrypoint.sh [hunk 27]
```diff
@@ -3,37 +3,41 @@
 set -e
 
 # Get MuJoCo key from the environment
-echo "${MJKEY}" > /root/.mujoco/mjkey.txt
+echo -e "${MJKEY}" > /root/.mujoco/mjkey.txt
```
Comment (ryanjulian) [inline docker/docker-entrypoint.sh]: why -e??

File: docker/docker-entrypoint.sh [hunk 28]
```diff
@@ -3,37 +3,41 @@
...
 # See: https://github.com/openai/mujoco-py/pull/145#issuecomment-356938564
-export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libGLEW.so:/usr/lib/x86_64-linux-gnu/mesa/libGL.so.1
+export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libGLEW.so
 
 export TF_CPP_MIN_LOG_LEVEL=3      # shut TensorFlow up
-export DISABLE_MUJOCO_RENDERING=1  # silence glfw
 export LD_LIBRARY_PATH="${LD_LIBRARY_PATH}:${HOME}/.mujoco/mjpro150/bin"
 
-exec "$@"
+source /root/.bashrc && exec "$@"
```
Comment (ryanjulian) [inline docker/docker-entrypoint.sh]: do not source a bashrc in a docker environment.

File: docker/docker-entrypoint.sh [hunk 29]
```diff
@@ -3,37 +3,41 @@
...
+  export DISABLE_MUJOCO_RENDERING=1  # silence glfw
 fi
 
 # Activate conda environment
 source activate garage
 
 # Fixes Segmentation Fault
 # See: https://github.com/openai/mujoco-py/pull/145#issuecomment-356938564
-export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libGLEW.so:/usr/lib/x86_64-linux-gnu/mesa/libGL.so.1
+export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libGLEW.so
```
Comment (ryanjulian) [inline docker/docker-entrypoint.sh]: does this work on nvidia? you would need to preload the nvidia version.

File: Makefile [hunk 30]
```diff
@@ -0,0 +1,43 @@
+.PHONY: build-garage-headless build-garage-nvidia \
```
Comment (ryanjulian) [inline Makefile]: i'm not sure this setup exactly works for interactive use cases. what will happen to all my data files if i'm using these containers to run an experiment?

by default, they will be destroyed when the container stops.

File: .travis.yml (hunk 15 above)
Comment (gonzaiva) [reply .travis.yml]: Removed

File: docker/Dockerfile (hunk 16 above)
Comment (gonzaiva) [reply docker/Dockerfile]: I realized that the docker-entrypoint for headless and nvidia shares the same commands except for creating the socket of the fake display for the headless type.
In order to distinguish between both I added this variable and I use it [here](https://github.com/rlworkgroup/garage/blob/b97856de130d0b5eebdfeb400796dc1eee458a78/docker/docker-entrypoint.sh#L8). My effort was to keep one canonical entry point for all image types, but now I see that the idea is that each type should have its own Dockerfile and entrypoint.

File: docker/Dockerfile (hunk 17 above)
Comment (gonzaiva) [reply docker/Dockerfile]: It is necessary in the headless (CI) version as well, but the config files is copied in [Travis](https://github.com/rlworkgroup/garage/blob/77714c38d5b575a5cfd6d1e42f0a045eebbe3484/.travis.yml#L28), and this won't work for local builds.
I forgot to remove the line from the .travis.yml though, so I will do that.

File: Makefile (hunk 18 above)
Comment (gonzaiva) [reply Makefile]: Done

File: docker/Dockerfile (hunk 19 above)
Comment (gonzaiva) [reply docker/Dockerfile]: Since I'm passing the MJKEY as a build argument in docker, I have to pass it as a single line string, that is, replace the new lines and returns as \n and \r, otherwise the make recipes are ruined by expanding these characters.
By using "echo -e", the backslash characters are expanded to reconstruct mjkey.txt.

File: docker/README.md (hunk 20 above)
Comment (gonzaiva) [reply docker/README.md]: I agree that is safer to pass the MJKEY file as a variable to the makefile, but the cat command is not enough because once MJKEY is expanded in the recipe, its first return/newline will indicate the end of the build command, and the rest of the build command will remain unprocessed, generating an error.
For example:
```
(garage) aigonzal@gonzaiva:~/ivanWorkspace/garage$ make build-garage-headless MJKEY="$(cat ~/.mujoco/mjkey.txt)"
docker build \
	-f docker/Dockerfile \
	--build-arg MJKEY="MuJoCo Pro Individual license activation key, number ????, type ?.
/bin/sh: 3: Syntax error: Unterminated quoted string
Makefile:14: recipe for target 'build-garage-headless' failed
make: *** [build-garage-headless] Error 2
```
That's why I'm using sed.

File: docker/README.md (hunk 21 above)
Comment (gonzaiva) [reply docker/README.md]: A difference between the headless and CI images could be CodeClimate, since only the latter uses it, but I don't see any problem in having CodeClimate for headless. I will remove the references to the CI image.

File: docker/README.md (hunk 22 above)
Comment (gonzaiva) [reply docker/README.md]: The user that only wants to test/deploy experiments with garage, sorry if the term is misleading. I will remove this section of text anyway.

File: docker/README.md (hunk 24 above)
Comment (gonzaiva) [reply docker/README.md]: I've checked docker-opengl, but it doesn't work when installing some apt dependencies. I'm still working on the opengl flavor.

File: docker/README.md (hunk 25 above)
Comment (gonzaiva) [reply docker/README.md]: I will replace the references from CI to headless.

File: docker/README.md (hunk 26 above)
Comment (gonzaiva) [reply docker/README.md]: ?

File: docker/docker-entrypoint.sh (hunk 27 above)
Comment (gonzaiva) [reply docker/docker-entrypoint.sh]: Since I'm passing the MJKEY as a build argument in docker, I have to pass it as a single line string, that is, replace the new lines and returns as \n and \r, otherwise the make recipes are ruined by expanding these characters.
By using "echo -e", the backslash characters are expanded to reconstruct mjkey.txt.

File: docker/docker-entrypoint.sh (hunk 29 above)
Comment (gonzaiva) [reply docker/docker-entrypoint.sh]: It does work on nvidia.

File: docker/docker-entrypoint.sh (hunk 28 above)
Comment (gonzaiva) [reply docker/docker-entrypoint.sh]: I'm sourcing bashrc because python packages used in the ROS workspace and the intera_interface are loaded in that way. Otherwise, if we try to run a python script that imports `intera_interface` and/or `rospy`, it will not find them.
A way to avoid this could be an environment variable with a list of all the setup scripts, and then source each of the scripts in the entrypoint, but that list has to be carried on from one docker build to another in dockercompose.
Any ideas to solve this problem are really appreciated :smiley_cat:

File: Makefile (hunk 30 above)
Comment (gonzaiva) [reply Makefile]: That's not completely true, I bind the data folder in the container with the data folder in the garage workspace using the argument:
```
-v $(shell pwd)/data:/root/code/garage/data
```
However, I accept this won't work when the container runs anywhere else in the file system. I will add a variable to pass the path to bind to the data folder in the container, having the garage/data folder as default, and if there's no folder in the path, then only in that case the data files will be lost.

**Comment by gonzaiva** [comment]
Variable expansion is not working when calling make in Travis as in my computer, and that's why the build fails. It may be related to using make 3.9.2 instead of 4.1 since the former comes installed with Ubuntu Trust, but I'm investigating other reasons.

File: docker/Dockerfile (hunk 17 above)
Comment (ryanjulian) [reply docker/Dockerfile]: what if i am an interactive user and i actually customized my config_personal.py though?

File: docker/Dockerfile.headless [hunk 31]
```diff
@@ -0,0 +1,9 @@
+ARG PARENT_IMAGE=rlworkgroup/garage-base
+FROM $PARENT_IMAGE
+
+# Pull CodeClimate reporter
```
Comment (ryanjulian) [inline docker/Dockerfile.headless]: perhaps you can do this in .travis.yml to avoid splitting the entrypoints? if the file is in code/garage it should get pulled into the image.

File: docker/README.md (hunk 26 above)
Comment (ryanjulian) [reply docker/README.md]: we aren't adding containers for the previous release, so this tag won't exist yet.

File: docker/docker-compose-headless.yml [hunk 32]
```diff
@@ -0,0 +1,14 @@
+version: '2.2'
```
Comment (ryanjulian) [inline docker/docker-compose-headless.yml]: is docker-compose really necessary here? we are only using one container at a time. docker-compose is for launching groups of containers which work together.

Comment (ryanjulian) [reply docker/docker-compose-headless.yml]: it's okay if it makes this more maintainable, but makefile+docker compose is starting to be a lot of layers.

File: docker/docker-entrypoint-headless.sh [hunk 33]
```diff
@@ -6,34 +6,36 @@ set -e
 echo "${MJKEY}" > /root/.mujoco/mjkey.txt
 
 # Setup dummy X server display
-display_num=0
+# Socket for display :0 may already be in use if the container is connected
+# to the network of the host, that's why we use 1.
```
Comment (ryanjulian) [inline docker/docker-entrypoint-headless.sh]: in that case you should start with a very high display number (e.g. 100) to avoid collisions with similar reasons.

File: docker/docker-entrypoint.sh (hunk 28 above)
Comment (ryanjulian) [reply docker/docker-entrypoint.sh]: that problem should be dealt with in garage-ros, not here. it's bad form to be sourcing bashrc inside a Docker container.

I realize ROS is full of bad design decisions, but we shouldn't let them percolate into these unrelated containers.

for instance, your ROS container can just be FROM this image, but with a different entrypoint.

File: docker/docker-entrypoint-nvidia.sh [hunk 34]
```diff
@@ -0,0 +1,18 @@
...
+source activate garage
+
+# Fixes Segmentation Fault
+# See: https://github.com/openai/mujoco-py/pull/145#issuecomment-356938564
+export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libGLEW.so
+
+export TF_CPP_MIN_LOG_LEVEL=3      # shut TensorFlow up
+export LD_LIBRARY_PATH="${LD_LIBRARY_PATH}:${HOME}/.mujoco/mjpro150/bin"
+
+source /root/.bashrc && exec "$@"
```
Comment (ryanjulian) [inline docker/docker-entrypoint-nvidia.sh]: please don't source bashrc inside docker.

File: Makefile [hunk 35]
```diff
@@ -0,0 +1,69 @@
...
+	run-garage-headless run-garage-nvidia
+
+# Path in host where the experiment data obtained in the container is stored
```
Comment (ryanjulian) [inline Makefile]: isn't this set in config.py/config_personal.py?

there needs to be a single source of truth here. you might need to ask Python where to put it.

File: docker/README.md (hunk 24 above)
Comment (ryanjulian) [reply docker/README.md]: okay. can you explain the difference between headless and CI then? i am still confused

File: Makefile (hunk 35 above)
Comment (gonzaiva) [reply Makefile]: Done

Comment (ryanjulian) [reply Makefile]: that solution (attempting to glob the conda install) is really hacky and makes me nervous. you have no guarantee that conda is even inside $HOME. i have machines where it lives in /opt, for instance. furthermore, you have no guarantee the host is running bash.

this is a good indication that at least one of the following things are true:
* Makefile is not a good way for us to manage the Dockerfiles (current config assumes Docker is launched from Python, so doesn't have this problem)
* We should not configure the user's environment in config.py

I think the least worst solution here is to pick a default place to map data/ from the docker container into (perhaps `$(pwd)/data/dockerfile_name` and stick to it. Allow users to override the default with an argument  to the Makefile

It also means we should remove the Docker-related entries from config/config_personal.py

It's clear here that how we do configuration in garage needs to be rewritten, but that can be left for another day.

File: Makefile (hunk 30 above)
Comment (ryanjulian) [reply Makefile]: in retrospect, something like this is actually the right thing to do

File: docker/Dockerfile (hunk 17 above)
Comment (ryanjulian) [reply docker/Dockerfile]: i think this logic needs to be moved to the Makefile

File: Makefile [hunk 36]
```diff
@@ -0,0 +1,80 @@
...
+		--build-arg MJKEY="$$(cat ${MJKEY_PATH})"
+
+build-garage-nvidia: docker/Dockerfile check-mjkey
+	docker-compose \
+		-f docker/docker-compose-nvidia.yml \
+		build \
+		--build-arg MJKEY="$$(cat ${MJKEY_PATH})"
+
+run-garage-ci: TAG ?= rlworkgroup/garage-ci
+run-garage-ci: docker/Dockerfile check-mjkey
```
Comment (ryanjulian) [inline Makefile]: docker/Dockerfile-headless?

File: Makefile (hunk 35 above)
Comment (gonzaiva) [reply Makefile]: For headless and nvidia I'm doing the following binding:
```
-v $(DATA_PATH)/$(CONTAINER_NAME):/root/code/garage/data
```
Where `DATA_PATH` is equal to `$(shell pwd)/data`
By default, `CONTAINER_NAME` for headless is equal to `garage-headless-$(DATE_TIME)`, and `garage-nvidia-$(DATE_TIME)` for nvidia. The DATE_TIME is defined by `$(shell date "+%Y_%m_%d_%H_%M_%S")`.
`CONTAINER_NAME` can be set by the user by passing it as an argument when calling make.

I created the target copy_config_personal to copy the config_personal.py file in case it doesn't exist.

File: .travis.yml [hunk 37]
```diff
@@ -24,49 +27,28 @@ before_install:
       "max-concurrent-uploads": 50
     }' | sudo tee /etc/docker/daemon.json
   - sudo service docker restart
-  # prevent garage from exiting on config
-  - cp garage/config_personal_template.py garage/config_personal.py
+  # Reinstall docker compose with the specified version
```
Comment (ryanjulian) [inline .travis.yml]: whoah this is strange. why can't we just pull the right version of docker from apt?

File: Makefile (hunk 35 above)
Comment (ryanjulian) [reply Makefile]: i think the path should be `$(shell pwd)/data/$(CONTAINER_NAME)`. the default experiment runner config already adds timecodes.

File: docker/Dockerfile [hunk 38]
```diff
@@ -99,13 +100,6 @@ WORKDIR /root/code/garage
 COPY .pre-commit-config.yaml /root/code/garage
 RUN ["/bin/bash", "-c", "source activate garage && git init && pre-commit"]
 
-# Pull CodeClimate reporter
-RUN curl -L https://codeclimate.com/downloads/test-reporter/test-reporter-latest-linux-amd64 > /root/code/garage/cc-test-reporter && \
-  chmod +x /root/code/garage/cc-test-reporter
-
 # Add code stub last
 COPY . /root/code/garage
 RUN ["/bin/bash", "-c", "source activate garage && pip install -e /root/code/garage && source deactivate garage"]
```
Comment (ryanjulian) [inline docker/Dockerfile]: the `pip install -e` step should be done in the Dockerfile, not the entrypoint

File: Makefile [hunk 39]
```diff
@@ -0,0 +1,87 @@
...
+# Date/time is appended to the container names
+DATE_TIME ?= $(shell date "+%Y_%m_%d_%H_%M_%S")
+# Prevent garage from exiting on config by making sure the personal config file
+# is already created
+CONFIG_PERSONAL := garage/config_personal.py
+
+build-garage-ci: TAG ?= rlworkgroup/garage-ci
+build-garage-ci: docker/docker-compose-headless.yml copy_config_personal
+	TAG=${TAG} \
+	docker-compose \
```
Comment (ryanjulian) [inline Makefile]: does this properly use `--cache-from`? otherwise, builds in the CI will take a very long time.

File: Makefile [hunk 40]
```diff
@@ -0,0 +1,87 @@
+.PHONY: build-garage-ci build-garage-headless build-garage-nvidia \
```
Comment (ryanjulian) [inline Makefile]: perhaps you can remove `-garage-` from all these directives, since it's obvious they are garage

`run-nvidia`, etc.

File: docker/Dockerfile (hunk 38 above)
Comment (gonzaiva) [reply docker/Dockerfile]: But this is the Dockerfile :confused:

File: Makefile [hunk 41]
```diff
@@ -0,0 +1,85 @@
...
+	run-garage-ci run-garage-headless run-garage-nvidia
+
+SHELL = bash
```
Comment (ryanjulian) [inline Makefile]: i don't think you have a guarantee bash is installed. is there a reason you need bash? i've basically never seen this in a Makefile before.

File: Makefile [hunk 42]
```diff
@@ -0,0 +1,85 @@
...
+
+check-mjkey:
+ifeq (0, $(shell [ ! -f $(MJKEY_PATH) ]; echo $$? ))
+	$(error The MJKEY was not set: make sure to pass a valid MJKEY_PATH \
+		or put your key at ~/.mujoco/mjkey.txt)
+endif
+
+copy_config_personal:
+ifeq (0, $(shell [ ! -f $(CONFIG_PERSONAL) ]; echo $$? ))
+	cp garage/config_personal_template.py garage/config_personal.py
```
Comment (ryanjulian) [inline Makefile]: you should warn a user when you do this

File: .travis.yml (hunk 37 above)
Comment (ryanjulian) [reply .travis.yml]: okay i see this in the Travis documentation.

can you include this link for explanation.

https://docs.travis-ci.com/user/docker/#using-docker-compose

File: Makefile (hunk 41 above)
Comment (gonzaiva) [reply Makefile]: bash is required to obtain the CodeCov arguments and pass them to the CI container:
```
run-ci:                                                                         
         docker run \                                                            
                 "$$(bash <(curl -s https://codecov.io/env))" \ 
                 ...
```
If the SHELL is not set as bash, even if the bash command is explicitly called in the target, the bourne shell is used. I can set the shell to be bash only in the run-ci target, for example:
```
run-ci: SHELL := /usr/bin/env bash
run-ci:                                                                         
         docker run \                                                            
                 "$$(bash <(curl -s https://codecov.io/env))" \ 
                 ...
```

File: .travis.yml (hunk 37 above)
Comment (gonzaiva) [reply .travis.yml]: Done

File: Makefile (hunk 40 above)
Comment (gonzaiva) [reply Makefile]: Done

File: docker/Dockerfile.headless [hunk 43]
```diff
@@ -0,0 +1,5 @@
+ARG PARENT_IMAGE=rlworkgroup/garage-base
+FROM $PARENT_IMAGE
+
+# Ready, set, go.
+ENTRYPOINT ["docker/docker-entrypoint-headless.sh"]
```
[the remark this answers is not in the document]
Comment (gonzaiva) [reply docker/Dockerfile.headless]: Done

File: Makefile (hunk 42 above)
Comment (gonzaiva) [reply Makefile]: DOne

File: Makefile (hunk 41 above)
Comment (gonzaiva) [reply Makefile]: I did that and worked, so let me know if this is okay with you, otherwise I can explore other options without using bash such as using a temporary file to save the script.

File: Makefile (hunk 39 above)
Comment (gonzaiva) [reply Makefile]: I have added the cache-from argument to the docker-compose files:
```

File: Makefile (hunk 41 above)
Comment (ryanjulian) [reply Makefile]: i like this solution. i think it might be more clear if you can put the results of the curl into a descriptive variable first (e.g. CI_ENV).

otherwise it would be hard for someone to figure out why we have a random curl call in our Docker run command.

Comment (gonzaiva) [reply Makefile]: Done
